compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -2321,6 +2321,13 @@ class Typer extends Namer
               val typer1 = localTyper(sym)
               typer1.typedDefDef(tree, sym)(using ctx.localContext(tree, sym).setTyper(typer1))
             case tree: untpd.TypeDef =>
+              if tree.name eq tpnme.? then
+                val addendum = if sym.owner.is(TypeParam)
```

Comment:
I'd move the whole case into a new method `typedTypeOrCassDef`, in order to keep the dispatch code in `typedUnnamed` as small as possible. That often helps the JOT to opimize.